biz.aQute.bndlib/src/aQute/bnd/osgi/MessageReporter.java

```diff
@@ -0,0 +1,279 @@
+
+		Cache() {
+			boolean failOk = processor().isFailOk();
+			Instructions fixupInstrs = new Instructions();
+			Parameters fixup = processor().getMergedParameters(Constants.FIXUPMESSAGES);
+			fixup.forEach((k, v) -> fixupInstrs.put(Instruction.legacy(k), v));
+
+			messages.values()
+				.stream()
+				.sorted()
+				.forEach(m -> {
+					m.fixup(fixupInstrs, failOk ? warnings : errors, warnings);
+				});
+		}
+	}
+
+	MessageReporter(Processor processor) {
+		this.processor = processor;
+	}
+
+	List<String> getWarnings() {
+		return fixup().warnings;
+	}
+
+	List<String> getErrors() {
+		return fixup().errors;
+	}
+
+	Cache fixup() {
+		return new Cache();
+	}
+
+	Location getLocation(String msg) {
+		return messages.get(msg);
+	}
+
+	public SetLocation error(String format, Object... args) {
+		try {
+			for (int i = 0; i < args.length; i++) {
+				if (args[i] instanceof Throwable t) {
+					args[i] = Exceptions.causes(t);
+				}
+			}
+			String s = Processor.formatArrays(format, args);
+			Message m = new Message(counter.getAndIncrement(), s, true);
+			putMessage(s, m);
+			return m;
+		} finally {
+			processor().signal();
+		}
+	}
+
+	void putMessage(String s, Message m) {
+		ConcurrentHashMap<String, Message> current;
+		do {
+			current = messages;
+			current.putIfAbsent(s, m);
+		} while (current != messages);
+	}
+
+	SetLocation warning(String format, Object... args) {
+		String s = Processor.formatArrays(format, args);
+		Message m = new Message(counter.getAndIncrement(), s, false);
+		putMessage(s, m);
+		return m;
+	}
+
+	void getInfo(Reporter from, String prefix) {
+		String actualPrefix = prefix == null ? prefix = processor().getBase() + " :" : prefix;
+
+		MessageReporter other;
+		if (from instanceof Processor processor) {
+			other = processor.reporter;
+		} else if (from instanceof MessageReporter mr) {
+			other = mr;
+		} else {
+			// backward compatible reporters
+			addAll(true, prefix, from);
+			addAll(false, prefix, from);
+			from.getErrors()
+				.clear();
+			from.getWarnings()
+				.clear();
+			return;
+		}
+
+		ConcurrentHashMap<String, Message> older = other.clear();
+		older.values()
+			.stream()
+			.sorted()
+			.map(m -> new Message(counter.getAndIncrement(), m, actualPrefix))
+			.forEach(m -> {
+				putMessage(m.message, m);
+			});
+	}
+
+	/*
+	 * for backward compatible reporters
+	 */
+	void addAll(boolean error, String prefix, Reporter reporter) {
+		for (String message : new ArrayList<>(error ? reporter.getErrors() : reporter.getWarnings())) {
+			String newMessage = prefix.isEmpty() ? message : prefix.concat(message);
+			Message m = new Message(counter.getAndIncrement(), newMessage, error);
+			Location location = reporter.getLocation(message);
+			if (location != null)
+				location.to(m);
+		}
+	}
+
+	ConcurrentHashMap<String, Message> clear() {
+		ConcurrentHashMap<String, Message> previous = messages;
+		messages = new ConcurrentHashMap<>();
```

Comment:
If you care, there is still a race condition here where errors can get lost.

Reply:
no, see the `putMessage()`. It is possible (but very unlikely) that the same message will be posted twice, which I think is acceptable.

To prevent this, it would not blocking between the `putMesssage()` and the `clear()` which I do not think is worth the overhead? But I am easy to convince here if you think that would be worth it?